packages/eslint-plugin/src/rules/prefer-enum-initializers.ts

```diff
@@ -1,41 +1,72 @@
 import { TSESTree } from '@typescript-eslint/experimental-utils';
 import * as util from '../util';
+import { RuleFix } from '@typescript-eslint/experimental-utils/dist/ts-eslint';
 
-export default util.createRule({
+type Option = '0-based' | '1-based' | 'key-name';
+type MessageIds = 'defineInitializer' | 'defineInitializerSuggestion';
+
+export default util.createRule<Option[], MessageIds>({
   name: 'prefer-enum-initializers',
   meta: {
     type: 'suggestion',
     docs: {
       description: 'Prefer initializing each enums member value',
       category: 'Best Practices',
       recommended: false,
+      suggestion: true,
     },
     messages: {
       defineInitializer:
         "The value of the member '{{ name }}' should be explicitly defined",
+      defineInitializerSuggestion:
+        'can be suggestion fixed to {{ name }} = {{ suggestedValue }}',
     },
-    schema: [],
+    schema: [
+      {
+        enum: ['0-based', '1-based', 'key-name'],
+      },
+    ],
   },
   defaultOptions: [],
   create(context) {
     const sourceCode = context.getSourceCode();
+    const config = context.options[0];
 
     return {
       TSEnumDeclaration(node: TSESTree.TSEnumDeclaration): void {
         const { members } = node;
-        const violatingMembers = members.filter(member => {
-          return member.initializer == null;
+        members.forEach((member, index) => {
+          if (member.initializer == null) {
+            const name = sourceCode.getText(member);
+            context.report({
+              node: member,
+              messageId: 'defineInitializer',
+              data: {
+                name,
+              },
+              suggest: [
+                {
+                  messageId: 'defineInitializerSuggestion',
+                  data: { name },
+                  fix: (fixer): RuleFix | null => {
+                    if (config === '0-based') {
+                      return fixer.replaceText(member, `${name} = ${index}`);
+                    } else if (config === '1-based') {
+                      return fixer.replaceText(
+                        member,
+                        `${name} = ${index + 1}`,
+                      );
+                    } else if (config === 'key-name') {
+                      return fixer.replaceText(member, `${name} = '${name}'`);
+                    } else {
+                      return null;
+                    }
+                  },
+                },
+              ],
```

Comment:
instead of only providing one suggestion and doing so behind an option, you could instead suggest all 3 variations at once!
`suggest` is an array of suggestions - and the IDE will display every single suggestion you provide!

This would give users the flexibility to pick and choose based on use-case, and would simplify the rule.

Example from the `no-explicit-any` rule, which suggests both fixing to `unknown` and to `never`
https://github.com/typescript-eslint/typescript-eslint/blob/817067b72b53c8f417f3c1c85e970ea82f0f6953/packages/eslint-plugin/src/rules/no-explicit-any.ts#L185-L198

Reply:
Thank you for your comment !
I'll fix that :bow:

I have a question about reporting.
Right now if there are several invalid members, `context.report` would be called for each member.
(https://github.com/typescript-eslint/typescript-eslint/pull/2326/files#diff-e35676a2d1a2277b5d1fab9a76471143R41)

Is it as you expected ?
I have tried to make it call once per an enum declaration but I have no idea how I make it :-(

Reply:
Every time you call `context.report`, it will report one error to eslint. ESLint does not do any aggregation for reports, it just shows the user every single individual report.
So because your implementation calls `context.report` once for each member without an initialiser, eslint will report an error per initialiser.

If you wanted to report just at the enum declaration, you'd have to instead do something like aggregate the list of broken members and then do a single `context.report` call with the `node` set to the `TSEnumDeclaration` node.

The issue with that is that you can only supply one location to the `context.report` then, so if there's one member without an initialiser, you'd just have to highlight the entire enum, which might be a bit too noisy.

This approach is good because it will highlight exactly the members that have issues, nothing more.